### ISSUE TITLE
Real EuclideanVector

### ISSUE THREAD

**Issue by davkovacs** [ACEsuit/ACE.jl#38]
Currently the ACE.EuclideanVector can only be of type complex, but I think it would make sense to have a real version too.

**Comment by cortner** [comment]
I agree - this is important, in fact it should be the default. @MatthiasSachs this is your job I'm afraid. Please think about how that ought to be done.

**Comment by davkovacs** [comment]
Maybe best to comment it here. 
I set up the basis and the fit the following way:
```
al_train = IPFitting.Data.read_xyz(joinpath(@__DIR__, "./water_test.xyz"),
                                  energy_key="", force_key="", dipole_key="dipole")

μ = ACE.EuclideanVector(Complex{Float64}); # site Dipole is an equivariant property

r0 = 1.1;
N = 3;

#degrees
Deg = ACE.TotalDegree(
      Dict( :n => 1.0, :l => 1.5 ),
      Dict( 1 => 1, 2 => 12/10, 3 => 12/8 )
      );

B1p = ACEatoms.ZμRnYlm_1pbasis(species = [:O, :H],
                               r0 = r0, rin = 0.5, rcut = 4.0, pin = 2,
                               D = Deg, maxdeg = 12);

pibasis1 = PIBasis(B1p, N, 4.0; property = μ, isreal=false);
ace_O = ACE.LinearACEModel(SymmetricBasis(pibasis1, μ));
ace_H = ACE.LinearACEModel(SymmetricBasis(pibasis1, μ));

Vsite = ACEatoms.ACESitePotential(Dict(:O => ace_O, :H => ace_H));

Bsite = ACEatoms.basis(Vsite);
@show length(Bsite);

B = Bsite
function JuLIP.MLIPs.combine(_B::typeof(B), coeffs::AbstractVector)
    @assert _B === B
    ACE.set_params!(Vsite, coeffs)
    return Vsite
end

dbname = ""
dB = LsqDB(dbname, Bsite, al_train);

weights = Dict("default" => Dict("MU" => 1.0 ));

IP_dip, lsqinfo = IPFitting.Lsq.lsqfit(dB, 
                                solver=(:lsqr, (0.001, 1e-7)),
                                asmerrs=true, weights=weights);
```

**Comment by davkovacs** [comment]
@MatthiasSachs Please let me know if you need any more information, I am afraid I have not managed to observe much more than this so far.

**Comment by cortner** [comment]
@MatthiasSachs Can you please start by reproducing David's results and then we need to debug this. I'm really surprised, I thought our tests were quite comprehensive.

**Comment by MatthiasSachs** [comment]
I am working on this. However, 
`al_train = IPFitting.Data.read_xyz(joinpath(@__DIR__, "./water_test.xyz"),
                                  energy_key="", force_key="", dipole_key="dipole")`
throws an error. I think the error is caused by the keyword `dipole_key="dipole" ` 
I am currently working on the dev branch of IPFitting.

**Comment by MatthiasSachs** [comment]
@cortner @davkovacs I guess David my have implemented that extension on his branch. Is there a way to quickly merge the relevant change into the developer branch of IPFitting?

**Comment by davkovacs** [comment]
Can you just checkout the dipole branch of IPFitting?

**Comment by MatthiasSachs** [comment]
@davkovacs could you send me the status of your julia environment / package manager ? I checked out the relevant dipole branches both for ACEatoms and IPFitting, but I am still having trouble getting things running...

**Comment by davkovacs** [comment]
Sorry about that, I have probably changed JuLIP as well to get this to work:
Project.toml:
```
[deps]
ACE = "3e8ccfd2-c8b0-11ea-32f1-f3a5990fd77a"
ACEatoms = "1e34e032-0b37-4839-a012-196b35408c3c"
ACEbase = "14bae519-eb20-449c-a949-9c58ed33163e"
IJulia = "7073ff75-c697-5162-941a-fcdaad2a7d2a"
IPFitting = "3002bd4c-79e4-52ce-b924-91256dde4e52"
IterativeSolvers = "42fd0dbc-a981-5370-80f2-aaf504508153"
JuLIP = "945c410c-986d-556a-acb1-167a618e0462"
NamedTupleTools = "d9ec5142-1e00-5aa0-9d6a-321866360f50"
StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
Zygote = "e88e6eb3-aa80-5325-afca-941959d7151f"
```

**Comment by MatthiasSachs** [comment]
oh, I should have been more precise: could you send me the output of 
] status
and the branches of all packages in development mode?

**Comment by davkovacs** [comment]
branches:
ACE: 'origin/dev-v0.10.x'.
ACEatoms: 'origin/dipole'.
IPFitting: 'origin/dipole'
JuLIP: 'origin/master'

**Comment by davkovacs** [comment]
I was wrong there, sorry nothing changed in JuLIP, I was planning to, but you warned me against it.

**Comment by cortner** [comment]
@MatthiasSachs Try the following configuration for the `Project.toml`:
```toml
[deps]
ACE = "3e8ccfd2-c8b0-11ea-32f1-f3a5990fd77a"
ACEatoms = "1e34e032-0b37-4839-a012-196b35408c3c"
ACEbase = "14bae519-eb20-449c-a949-9c58ed33163e"
IJulia = "7073ff75-c697-5162-941a-fcdaad2a7d2a"
IPFitting = "3002bd4c-79e4-52ce-b924-91256dde4e52"
IterativeSolvers = "42fd0dbc-a981-5370-80f2-aaf504508153"
JuLIP = "945c410c-986d-556a-acb1-167a618e0462"
NamedTupleTools = "d9ec5142-1e00-5aa0-9d6a-321866360f50"
StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
Zygote = "e88e6eb3-aa80-5325-afca-941959d7151f"

[compat]
ACE = "0.10"
ACEbase = "0.1.5"
julia = "1.5,1.6"
[deps]
ACE = "3e8ccfd2-c8b0-11ea-32f1-f3a5990fd77a"
ACEatoms = "1e34e032-0b37-4839-a012-196b35408c3c"
ACEbase = "14bae519-eb20-449c-a949-9c58ed33163e"
IJulia = "7073ff75-c697-5162-941a-fcdaad2a7d2a"
IPFitting = "3002bd4c-79e4-52ce-b924-91256dde4e52"
IterativeSolvers = "42fd0dbc-a981-5370-80f2-aaf504508153"
JuLIP = "945c410c-986d-556a-acb1-167a618e0462"
NamedTupleTools = "d9ec5142-1e00-5aa0-9d6a-321866360f50"
StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
Zygote = "e88e6eb3-aa80-5325-afca-941959d7151f"

[compat]
ACE = "0.10"
ACEbase = "0.1.5"
julia = "1.5,1.6"
```

**Comment by cortner** [comment]
and then `dev` `IPFitting` and `ACEatoms` both on the `dipole` branches

**Comment by davkovacs** [comment]
I think we have solved this @MatthiasSachs I was putting the `isreal=true` to the PIbasis and not the SymmetricBasis, sorry about that. All works now the same irrespective of putting real or complex there.

**Comment by MatthiasSachs** [comment]
Ok, great! @cortner Would it make sense to output a warning if an inconsistent combination of the `isreal` flags is used in SymmetricBasis and PIBasis?

**Comment by cortner** [comment]
I'd like to come back to this with the hindsight of a bit more experience. This is an interesting design issue actually. 

If you give me a property are you able to tell me whether there should be a `real` call between the product basis and the symmetrisation, and another `real` call after symmetrisation? Or would the user have to specify that manually? 

IOW, I'd like to make this fully automated, allowing the `property` to auto-magically insert the call to `real` before and/or after the symmetrisation is applied. The main design question is whether the product basis needs to know about it, or not. This would save a little bit of performance especially at low correlation order but I'm tempted to just ignore that.

**Comment by cortner** [comment]
As I'm thinking about it, I think I want to suggest the following: 
```julia
pre_a2b(aa, property)      # apply this before the A2B mapping 
post_a2b(aa, property)    # apply this after the A2B mapping 
```
maybe there are better names...

**Comment by cortner** [comment]
This should be fully resolved with #51
